### PR TITLE
feature/issue-136-connect-bidding-system-to-backend

### PR DESCRIPTION
**Pull request by NervzZ**
Finished the bidding system implementations such that :
- We can now properly create bidding listings
- We can place bids on those
- If we currently hold the best bid, it is stated on the listings details, if not we can access the bid controls like any other users.

### CODE REVIEW

**Comment by NervzZ** [comment]
Tests incoming

File: app/src/main/java/com/studhub/app/presentation/listing/details/DetailedListingViewModel.kt [hunk 1]
```diff
@@ -98,6 +105,39 @@ class DetailedListingViewModel @Inject constructor(
...
+     */
+    override fun placeBid(bid: Float?, onError: (msg: String) -> Unit) {
+        if (bid == null) {
+            onError("Your bid is not a valid value")
+        } else {
+            when (currentListing) {
+                is ApiResponse.Success -> {
+                    val listing = (currentListing as ApiResponse.Success<Listing>).data
+                    if (bid <= listing.price) {
+                        onError("Your bid needs to be greater than the current one")
```
Comment (tepoorte) [inline app/src/main/java/com/studhub/app/presentation/listing/details/DetailedListingViewModel.kt:122]: Type for currentListing should be explicit, I would avoid using type inference.

File: app/src/main/java/com/studhub/app/presentation/listing/details/DetailedListingViewModel.kt [hunk 2]
```diff
@@ -98,6 +105,39 @@ class DetailedListingViewModel @Inject constructor(
...
+                    } else {
+                        viewModelScope.launch {
+                            //keeps only 2 decimals after floating point
+                            placeBid(listing, String.format("%.2f", bid).toFloat()).collect {
+                                when (it) {
+                                    is ApiResponse.Loading -> {}
```
Comment (tepoorte) [inline app/src/main/java/com/studhub/app/presentation/listing/details/DetailedListingViewModel.kt:128]: Well done for String.format !

File: app/src/main/java/com/studhub/app/presentation/listing/details/components/BiddingControls.kt [hunk 3]
```diff
@@ -14,21 +16,53 @@ import com.studhub.app.R
...
+fun BiddingControls(
+    price: MutableState<String>,
+    deadline: Date,
+    onSubmit: () -> Unit = {},
+    hasBid: Boolean
+) {
+    val calendar = GregorianCalendar()
+    calendar.time = deadline
+    val formatter = SimpleDateFormat("dd.MM.yyyy")
+    Column(
```
Comment (tepoorte) [inline app/src/main/java/com/studhub/app/presentation/listing/details/components/BiddingControls.kt:43]: pay attention for the date format, In the createListing interface we actually have _MM.DD.YYYY_

**Review by tepoorte — CHANGES_REQUESTED** [review]
Nice work, well done! Just some very little improvements before merging can be done.

File: app/src/main/java/com/studhub/app/presentation/listing/details/components/BiddingControls.kt (hunk 3 above)
Comment (NervzZ) [reply app/src/main/java/com/studhub/app/presentation/listing/details/components/BiddingControls.kt:43]: Doesn't really matter, that's the default display format on top of the datepicker but we can see the input is in the usual format I'm using to display the date :
image

File: app/src/androidTest/java/com/studhub/app/presentation/listing/details/DetailedListingViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    @Test
+    fun detailedListingViewModelContactSellerSucceeds() {
+        var callback = false
+        runBlocking {
+            viewModel.contactSeller(seller = seller, callback = {callback = true})
+            delay(100)
+        }
+        assert(callback)
+    }
```
[the remark this answers is not in the document]
Comment (NervzZ) [reply app/src/androidTest/java/com/studhub/app/presentation/listing/details/DetailedListingViewModelTest.kt:192]: good catch

File: app/src/androidTest/java/com/studhub/app/presentation/listing/details/DetailedListingScreenMainComposableTest.kt [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    private val startConversationWith = StartConversationWith(convoRepo, authRepo)
+    private val addBlockedUser = AddBlockedUser(userRepo, authRepo)
+    private val unblockUser = UnblockUser(userRepo, authRepo)
+    private val getBlockedUsers = GetBlockedUsers(userRepo, authRepo)
+    private val auth = authRepo
+    private val placeBid = PlaceBid(listingRepo, authRepo)
+
+    private val viewModel = DetailedListingViewModel(
+        getListing,
+        getFavoriteListings,
```
[the remark this answers is not in the document]
Comment (NervzZ) [reply app/src/androidTest/java/com/studhub/app/presentation/listing/details/DetailedListingScreenMainComposableTest.kt:50]: It's the only viewmodel of the file and it's a testing file about DetailedListingScreen.

**Pull request merged**